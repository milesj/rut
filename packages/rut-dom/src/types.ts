/* eslint-disable @typescript-eslint/no-namespace */

import React from 'react';
import { ElementType } from 'rut';
import DomElement from './DomElement';

export type HostComponentType = keyof JSX.IntrinsicElements;

export type InferHostElement<T> = T extends keyof HTMLElementTagNameMap
  ? HTMLElementTagNameMap[T]
  : T extends keyof SVGElementTagNameMap
  ? SVGElementTagNameMap[T]
  : unknown;

export type InferComponentProps<T> = T extends HostComponentType
  ? JSX.IntrinsicElements[T]
  : T extends React.ComponentType<infer P>
  ? P
  : never;

// EVENTS

export type EventMap<T> = Required<
  Omit<React.DOMAttributes<T>, 'children' | 'dangerouslySetInnerHTML'>
>;

export type EventType = keyof EventMap<unknown>;

export type EventOptions<T, E> = {
  currentTarget?: Partial<T>;
  target?: Partial<T>;
} & ExpandedEventOptions<E>;

export type ExpandedEventOptions<T> = T extends React.AnimationEvent | AnimationEvent
  ? { animationName?: string }
  : T extends
      | React.MouseEvent
      | React.KeyboardEvent
      | React.TouchEvent
      | MouseEvent
      | KeyboardEvent
      | TouchEvent
  ? {
      altKey?: boolean;
      ctrlKey?: boolean;
      key?: string;
      keyCode?: number;
      metaKey?: boolean;
      shiftKey?: boolean;
    }
  : T extends React.TransitionEvent | TransitionEvent
  ? { propertyName?: string }
  : {};

export type DomPropsOf<T> = T extends DomElement<ElementType, infer P> ? P : never;

declare global {
  namespace jest {
    interface Matchers<R, T> {
      toHaveProp<K extends keyof DomPropsOf<T>>(name: K, value?: DomPropsOf<T>[K]): R;
      toHaveProps(props: Partial<DomPropsOf<T>>): R;
    }
  }
}
