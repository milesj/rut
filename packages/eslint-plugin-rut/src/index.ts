/**
 * @copyright   2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import rules from './rules';

export = {
  rules,
  configs: {
    recommended: {
      plugins: ['rut'],
      rules: {
        'rut/no-act': 'error',
      },
    },
  },
};
