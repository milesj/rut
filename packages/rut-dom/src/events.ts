// https://developer.mozilla.org/en-US/docs/Web/Events
// istanbul ignore next
export function createEvent(type: string): Event {
  switch (type) {
    case 'animationcancel':
    case 'animationiteration':
    case 'animationend':
    case 'animationstart':
      return new AnimationEvent(type);
    case 'beforeunload':
      return new BeforeUnloadEvent();
    case 'auxclick':
    case 'click':
    case 'contextmenu':
    case 'dblclick':
    case 'mousedown':
    case 'mouseenter':
    case 'mouseleave':
    case 'mousemove':
    case 'mouseout':
    case 'mouseover':
    case 'mouseup':
      return new MouseEvent(type);
    case 'devicelight':
    case 'devicemotion':
    case 'deviceorientation':
    case 'deviceorientationabsolute':
      return new DeviceLightEvent(type);
    case 'blur':
    case 'focus':
      return new FocusEvent(type);
    case 'copy':
    case 'cut':
    case 'paste':
      return new ClipboardEvent(type);
    case 'compositionend':
    case 'compositionstart':
    case 'compositionupdate':
      return new CompositionEvent(type);
    case 'drag':
    case 'dragend':
    case 'dragenter':
    case 'dragexit':
    case 'dragleave':
    case 'dragover':
    case 'dragstart':
    case 'drop':
      return new DragEvent(type);
    case 'error':
      return new ErrorEvent(type);
    case 'hashchange':
      return new HashChangeEvent(type);
    case 'keydown':
    case 'keypress':
    case 'keyup':
      return new KeyboardEvent(type);
    case 'loadend':
    case 'progress':
    case 'readystatechange':
      return new ProgressEvent(type);
    case 'message':
    case 'messageerror':
      return new MessageEvent(type);
    case 'pagehide':
    case 'pageshow':
      return new PageTransitionEvent();
    case 'gotpointercapture':
    case 'lostpointercapture':
    case 'pointercancel':
    case 'pointerdown':
    case 'pointerenter':
    case 'pointerleave':
    case 'pointerover':
    case 'pointermove':
    case 'pointerout':
    case 'pointerup':
      return new PointerEvent(type);
    case 'popstate':
      return new PopStateEvent(type);
    case 'touchcancel':
    case 'touchend':
    case 'touchmove':
    case 'touchstart':
      return new TouchEvent(type);
    case 'transitioncancel':
    case 'transitionend':
    case 'transitionrun':
    case 'transitionstart':
      return new TransitionEvent(type);
    case 'abort':
    case 'scroll':
    case 'resize':
      return new UIEvent(type);
    case 'securitypolicyviolation':
      return new SecurityPolicyViolationEvent(type);
    case 'storage':
      return new StorageEvent(type);
    case 'unhandledrejection':
      return new PromiseRejectionEvent(type, { promise: Promise.resolve() });
    case 'wheel':
      return new WheelEvent(type);
    default:
      return new Event(type);
  }
}
