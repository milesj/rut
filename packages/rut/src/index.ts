/**
 * @copyright   2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { act } from 'react-test-renderer';
import matchers from './matchers';
import { render, renderAndWait } from './render';
import mockSyntheticEvent from './mocks/event';

export * from './helpers';
export * from './types';

export { act, matchers, mockSyntheticEvent, render, renderAndWait };
