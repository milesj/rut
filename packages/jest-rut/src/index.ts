/**
 * @copyright   2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { unmockFetches, MatchResult, integrate, matchers } from 'rut';
import { runWithTimers } from './integration';
import serializer from './serializer';

integrate({
  runWithTimers,
});

afterEach(unmockFetches);

const jestMatchers: jest.ExpectExtendMap = {};

Object.entries(matchers).forEach(([name, matcher]) => {
  jestMatchers[name] = function jestMatcher(this: jest.MatcherUtils, ...args: unknown[]) {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    const result: MatchResult = (matcher as Function).call(this, ...args);
    let message = `${this.utils.matcherHint(result.name)}\n\n`;

    // Append error message
    message += (this.isNot ? result.notMessage : result.message)
      .replace('{{received}}', this.utils.printReceived(result.received))
      .replace('{{expected}}', result.expected ? this.utils.printExpected(result.expected) : '')
      .replace('{{actual}}', result.actual ? String(result.actual) : '');

    // Append diff if requested
    if (result.diff) {
      message += '\n';
      message += this.utils.diff(result.expected, result.actual, {
        aAnnotation: 'Expected',
        bAnnotation: 'Actual',
      });
    }

    return {
      message: () => message,
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      pass: typeof result.passed === 'function' ? result.passed(this.equals) : result.passed,
    };
  };
});

expect.extend(jestMatchers);
expect.addSnapshotSerializer(serializer);

export default jestMatchers;
