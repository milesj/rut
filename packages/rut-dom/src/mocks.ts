/* eslint-disable complexity */

import React from 'react';
import { InferHostElementFromEvent } from 'rut';
import { BaseEvent, SyntheticEvent } from 'rut/lib/adapters';
import { EventType, EventOptions } from './types';

/**
 * In both the `mockEvent` and `mockSyntheticEvent` functions below,
 * you'll notice that every return is being type ignored.
 * We are doing this so that the generic type takes precendence,
 * as we want inferrence to be predominantly used.
 *
 * For example, if a test is dispatching `findOne('button').dispatch('onClick`)`,
 * the first argument is typed as `React.MouseEvent<HTMLButtonElement, MouseEvent>`.
 * We can take advantage of type inferrence by mocking the argument at
 * the call site, like so: `findOne('button').dispatch('onClick', mockSyntheticEvent('onClick'))`.
 *
 * With this pattern, mocks are easily typed, and the underlying event object
 * structure is close enough for most, if not all of test cases.
 */

// https://developer.mozilla.org/en-US/docs/Web/Events
// istanbul ignore next
function createEvent(type: string): Event {
  switch (type) {
    case 'animationcancel':
    case 'animationiteration':
    case 'animationend':
    case 'animationstart':
      return new AnimationEvent(type);
    case 'beforeunload':
      return new BeforeUnloadEvent();
    case 'auxclick':
    case 'click':
    case 'contextmenu':
    case 'dblclick':
    case 'mousedown':
    case 'mouseenter':
    case 'mouseleave':
    case 'mousemove':
    case 'mouseout':
    case 'mouseover':
    case 'mouseup':
      return new MouseEvent(type);
    case 'devicelight':
    case 'devicemotion':
    case 'deviceorientation':
    case 'deviceorientationabsolute':
      return new DeviceLightEvent(type);
    case 'blur':
    case 'focus':
      return new FocusEvent(type);
    case 'copy':
    case 'cut':
    case 'paste':
      return new ClipboardEvent(type);
    case 'compositionend':
    case 'compositionstart':
    case 'compositionupdate':
      return new CompositionEvent(type);
    case 'drag':
    case 'dragend':
    case 'dragenter':
    case 'dragexit':
    case 'dragleave':
    case 'dragover':
    case 'dragstart':
    case 'drop':
      return new DragEvent(type);
    case 'error':
      return new ErrorEvent(type);
    case 'hashchange':
      return new HashChangeEvent(type);
    case 'keydown':
    case 'keypress':
    case 'keyup':
      return new KeyboardEvent(type);
    case 'loadend':
    case 'progress':
    case 'readystatechange':
      return new ProgressEvent(type);
    case 'message':
    case 'messageerror':
      return new MessageEvent(type);
    case 'pagehide':
    case 'pageshow':
      return new PageTransitionEvent(type);
    case 'gotpointercapture':
    case 'lostpointercapture':
    case 'pointercancel':
    case 'pointerdown':
    case 'pointerenter':
    case 'pointerleave':
    case 'pointerover':
    case 'pointermove':
    case 'pointerout':
    case 'pointerup':
      return new PointerEvent(type);
    case 'popstate':
      return new PopStateEvent(type);
    case 'touchcancel':
    case 'touchend':
    case 'touchmove':
    case 'touchstart':
      return new TouchEvent(type);
    case 'transitioncancel':
    case 'transitionend':
    case 'transitionrun':
    case 'transitionstart':
      return new TransitionEvent(type);
    case 'abort':
    case 'scroll':
    case 'resize':
      return new UIEvent(type);
    case 'securitypolicyviolation':
      return new SecurityPolicyViolationEvent(type);
    case 'storage':
      return new StorageEvent(type);
    case 'unhandledrejection':
      return new PromiseRejectionEvent(type, { promise: Promise.resolve() });
    case 'wheel':
      return new WheelEvent(type);
    default:
      return new Event(type);
  }
}

/**
 * Mock a DOM `Event` based on type.
 */
export function mockEvent<T = Event>(type: string, options?: EventOptions<Element, T>): T {
  const { currentTarget, target, ...props } = options || {};
  let event: Event;

  // JSDOM environment does not exist, which means we do not have events.
  // Return a very custom low-level event object for the time being.
  if (typeof window === 'undefined') {
    // @ts-ignore Ignore legacy fields
    event = new BaseEvent(type);
  } else {
    event = createEvent(type);
  }

  if (target) {
    Object.defineProperty(event, 'target', {
      enumerable: true,
      value: target,
    });
  }

  if (currentTarget || target) {
    Object.defineProperty(event, 'currentTarget', {
      enumerable: true,
      value: currentTarget || target,
    });
  }

  Object.entries(props).forEach(([prop, value]) => {
    Object.defineProperty(event, prop, {
      enumerable: true,
      value,
    });
  });

  // @ts-ignore Infer generics
  return event;
}

/**
 * Mock a React `SyntheticEvent` based on type.
 */
export function mockSyntheticEvent<T = React.SyntheticEvent>(
  type: EventType,
  options?: EventOptions<InferHostElementFromEvent<T>, T>,
): T {
  let eventType = type.toLowerCase();

  if (eventType.startsWith('on')) {
    eventType = eventType.slice(2);
  }

  // @ts-ignore Infer generics
  return new SyntheticEvent(eventType, mockEvent(eventType, options));
}
