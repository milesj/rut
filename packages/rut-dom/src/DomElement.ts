/* eslint-disable lines-between-class-members, no-dupe-class-members, @typescript-eslint/no-explicit-any */

import React from 'react';
import { AtIndexType, DispatchOptions, InferEventFromHandler, UnknownProps } from 'rut';
import { Element, SyntheticEvent } from 'rut/lib/adapters';
import { mockSyntheticEvent } from './mocks';
import {
  HostComponentType,
  InferComponentProps,
  InferHostElement,
  EventMap,
  EventType,
  EventOptions,
} from './types';

export default class DomElement<
  Type extends React.ElementType = React.ElementType,
  Props = InferComponentProps<Type>,
  Host = InferHostElement<Type>
> extends Element<Type, Props, Host> {
  createSyntheticEvent(
    eventType: EventType,
    event: unknown,
    elementType: React.ElementType,
  ): React.SyntheticEvent {
    // Event provided by consumer, so use as is
    if (event instanceof SyntheticEvent) {
      return event as React.SyntheticEvent;
    }

    // Either event options or nothing provided
    const options: EventOptions<HTMLElement, Event> =
      typeof event === 'object' && event !== null ? event : {};

    // Set a target automatically if not provided
    if (typeof elementType === 'string' && !options.target) {
      if (typeof document === 'undefined') {
        options.target = { tagName: elementType.toUpperCase() };
      } else {
        options.target = document.createElement(elementType);
      }
    }

    return mockSyntheticEvent(eventType, options);
  }

  dispatch<K extends EventType>(
    name: K,
    eventOrConfig?:
      | InferEventFromHandler<EventMap<Host>[K]>
      | EventOptions<Host, InferEventFromHandler<EventMap<Host>[K]>>,
    options: DispatchOptions = {},
  ): this {
    return super.dispatch(name, eventOrConfig, options);
  }

  async dispatchAndWait<K extends EventType>(
    name: K,
    eventOrConfig?:
      | InferEventFromHandler<EventMap<Host>[K]>
      | EventOptions<Host, InferEventFromHandler<EventMap<Host>[K]>>,
    options: DispatchOptions = {},
  ): Promise<void> {
    return super.dispatchAndWait(name, eventOrConfig, options);
  }

  find<T extends HostComponentType, P extends InferComponentProps<T>>(
    type: T,
    props?: Partial<P>,
  ): DomElement<T, P>[];
  find<T extends React.ComponentType<any>, P extends InferComponentProps<T>>(
    type: T,
    props?: Partial<P>,
  ): DomElement<T, P>[];
  find(type: React.ElementType<unknown>, props?: UnknownProps): DomElement<React.ElementType>[] {
    return super.find(type, props);
  }

  findAt<T extends HostComponentType, P extends InferComponentProps<T>>(
    type: T,
    at: AtIndexType,
    props?: Partial<P>,
  ): DomElement<T, P>;
  findAt<T extends React.ComponentType<any>, P extends InferComponentProps<T>>(
    type: T,
    at: AtIndexType,
    props?: Partial<P>,
  ): DomElement<T, P>;
  findAt(
    type: React.ElementType<unknown>,
    at: AtIndexType,
    props?: UnknownProps,
  ): DomElement<React.ElementType> {
    return super.findAt(type, at, props);
  }

  findOne<T extends HostComponentType, P extends InferComponentProps<T>>(
    type: T,
    props?: Partial<P>,
  ): DomElement<T, P>;
  findOne<T extends React.ComponentType<any>, P extends InferComponentProps<T>>(
    type: T,
    props?: Partial<P>,
  ): DomElement<T, P>;
  findOne(type: React.ElementType<unknown>, props?: UnknownProps): DomElement<React.ElementType> {
    return super.findOne(type, props);
  }
}
