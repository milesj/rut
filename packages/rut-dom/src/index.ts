/**
 * @copyright   2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { RendererOptions } from 'rut';
import { doRender, doRenderAndWait } from 'rut/lib/adapters';
import DomElement from './DomElement';

export * from 'rut';
export * from './mocks';
export * from './types';

export function render<Props extends object = {}>(
  element: React.ReactElement,
  options?: RendererOptions,
) {
  return doRender<Props, DomElement<React.ComponentType<Props>, Props>>(element, {
    ...options,
    createElement: instance => new DomElement(instance),
  });
}

export function renderAndWait<Props extends object = {}>(
  element: React.ReactElement,
  options?: RendererOptions,
) {
  return doRenderAndWait<Props, DomElement<React.ComponentType<Props>, Props>>(element, {
    ...options,
    createElement: instance => new DomElement(instance),
  });
}
