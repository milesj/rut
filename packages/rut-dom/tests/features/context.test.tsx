import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import { render } from '../../src';

describe('Context', () => {
  const ThemeContext = React.createContext('default');

  interface SpyProps {
    spy: (theme: string) => void;
  }

  interface ProviderProps {
    children: React.ReactNode;
    theme?: string;
  }

  function Provider({ children, theme }: ProviderProps) {
    return <ThemeContext.Provider value={theme || 'light'}>{children}</ThemeContext.Provider>;
  }

  function Consumer({ spy }: SpyProps) {
    return (
      <ThemeContext.Consumer>
        {theme => {
          spy(theme);

          return <div>Consumed!</div>;
        }}
      </ThemeContext.Consumer>
    );
  }

  function HookConsumer({ spy }: SpyProps) {
    const theme = useContext(ThemeContext);

    spy(theme);

    return <div>Consumed!</div>;
  }

  class StaticConsumer extends React.Component<SpyProps> {
    static contextType = ThemeContext;

    render() {
      this.props.spy(this.context);

      return <div>Consumed!</div>;
    }
  }

  describe('Provider', () => {
    it('renders a provider without consumer, and its children', () => {
      const result = render<ProviderProps>(<Provider>Child</Provider>);

      expect(result).toMatchSnapshot();
      expect(result.root).toContainNode('Child');
    });
  });

  describe('Consumer', () => {
    it('renders without provider', () => {
      const spy = jest.fn();

      render<SpyProps>(<Consumer spy={spy} />);

      expect(spy).toHaveBeenCalledWith('default');
    });

    it('renders within provider and inherits the defined value', () => {
      const spy = jest.fn();
      const result = render<ProviderProps>(
        <Provider theme="dark">
          <div>
            <Consumer spy={spy} />
          </div>
        </Provider>,
      );

      expect(result).toMatchSnapshot();
      expect(spy).toHaveBeenCalledWith('dark');
    });

    it('updates with new value when provider changes', () => {
      const spy = jest.fn();
      const result = render<ProviderProps>(
        <Provider theme="dark">
          <div>
            <Consumer spy={spy} />
          </div>
        </Provider>,
      );

      expect(spy).toHaveBeenCalledWith('dark');

      result.update({ theme: 'light' });

      expect(spy).toHaveBeenCalledWith('light');
    });

    it('updates consumer that has been memoized', () => {
      const spy = jest.fn();
      const MemoConsumer = React.memo(Consumer);
      const result = render<ProviderProps>(
        <Provider theme="dark">
          <MemoConsumer spy={spy} />
        </Provider>,
      );

      result.update({ theme: 'light' });

      expect(spy).toHaveBeenCalledTimes(2);
      expect(spy).toHaveBeenCalledWith('light');
    });
  });

  describe('useContext', () => {
    it('renders without provider', () => {
      const spy = jest.fn();

      render<SpyProps>(<HookConsumer spy={spy} />);

      expect(spy).toHaveBeenCalledWith('default');
    });

    it('renders within provider and inherits the defined value', () => {
      const spy = jest.fn();
      const result = render<ProviderProps>(
        <Provider theme="dark">
          <div>
            <HookConsumer spy={spy} />
          </div>
        </Provider>,
      );

      expect(result).toMatchSnapshot();
      expect(spy).toHaveBeenCalledWith('dark');
    });

    it('updates with new value when provider changes', () => {
      const spy = jest.fn();
      const result = render<ProviderProps>(
        <Provider theme="dark">
          <div>
            <HookConsumer spy={spy} />
          </div>
        </Provider>,
      );

      expect(spy).toHaveBeenCalledWith('dark');

      result.update({ theme: 'light' });

      expect(spy).toHaveBeenCalledWith('light');
    });

    it('updates consumer that has been memoized', () => {
      const spy = jest.fn();
      const MemoConsumer = React.memo(HookConsumer);
      const result = render<ProviderProps>(
        <Provider theme="dark">
          <MemoConsumer spy={spy} />
        </Provider>,
      );

      result.update({ theme: 'light' });

      expect(spy).toHaveBeenCalledTimes(2);
      expect(spy).toHaveBeenCalledWith('light');
    });
  });

  describe('contextType', () => {
    it('renders without provider', () => {
      const spy = jest.fn();

      render<SpyProps>(<StaticConsumer spy={spy} />);

      expect(spy).toHaveBeenCalledWith('default');
    });

    it('renders within provider and inherits the defined value', () => {
      const spy = jest.fn();
      const result = render<ProviderProps>(
        <Provider theme="dark">
          <div>
            <StaticConsumer spy={spy} />
          </div>
        </Provider>,
      );

      expect(result).toMatchSnapshot();
      expect(spy).toHaveBeenCalledWith('dark');
    });

    it('updates with new value when provider changes', () => {
      const spy = jest.fn();
      const result = render<ProviderProps>(
        <Provider theme="dark">
          <div>
            <StaticConsumer spy={spy} />
          </div>
        </Provider>,
      );

      expect(spy).toHaveBeenCalledWith('dark');

      result.update({ theme: 'light' });

      expect(spy).toHaveBeenCalledWith('light');
    });

    it('updates consumer that is pure', () => {
      class PureStaticConsumer extends React.PureComponent<SpyProps> {
        static contextType = ThemeContext;

        render() {
          this.props.spy(this.context);

          return <div>Consumed!</div>;
        }
      }

      const spy = jest.fn();
      const result = render<ProviderProps>(
        <Provider theme="dark">
          <PureStaticConsumer spy={spy} />
        </Provider>,
      );

      result.update({ theme: 'light' });

      expect(spy).toHaveBeenCalledTimes(2);
      expect(spy).toHaveBeenCalledWith('light');
    });
  });

  describe('legacy', () => {
    class ParentContext extends React.Component<{ children: React.ReactNode; theme?: string }> {
      static childContextTypes = {
        theme: PropTypes.string,
      };

      getChildContext() {
        return { theme: this.props.theme || 'light' };
      }

      render() {
        // eslint-disable-next-line react/jsx-no-useless-fragment
        return <>{this.props.children}</>;
      }
    }

    class ChildContext extends React.Component<SpyProps> {
      static contextTypes = {
        theme: PropTypes.string,
      };

      render() {
        this.props.spy(this.context.theme);

        return <div>Consumed!</div>;
      }
    }

    it('renders without provider', () => {
      const spy = jest.fn();

      render<SpyProps>(<ChildContext spy={spy} />);

      expect(spy).toHaveBeenCalledWith(undefined);
    });

    it('renders within parent and inherits the defined value', () => {
      const spy = jest.fn();
      const result = render<ProviderProps>(
        <ParentContext theme="dark">
          <div>
            <ChildContext spy={spy} />
          </div>
        </ParentContext>,
      );

      expect(result).toMatchSnapshot();
      expect(spy).toHaveBeenCalledWith('dark');
    });

    it('updates with new value when provider changes', () => {
      const spy = jest.fn();
      const result = render<ProviderProps>(
        <ParentContext theme="dark">
          <div>
            <ChildContext spy={spy} />
          </div>
        </ParentContext>,
      );

      expect(spy).toHaveBeenCalledWith('dark');

      result.update({ theme: 'light' });

      expect(spy).toHaveBeenCalledWith('light');
    });
  });
});
