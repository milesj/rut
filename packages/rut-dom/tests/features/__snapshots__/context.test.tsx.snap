// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Context Consumer renders within provider and inherits the defined value 1`] = `
<Provider theme="dark">
  <div>
    <Consumer spy={mockConstructor()}>
      <div>Consumed!</div>
    </Consumer>
  </div>
</Provider>
`;

exports[`Context Provider renders a provider without consumer, and its children 1`] = `<Provider>Child</Provider>`;

exports[`Context contextType renders within provider and inherits the defined value 1`] = `
<Provider theme="dark">
  <div>
    <StaticConsumer spy={mockConstructor()}>
      <div>Consumed!</div>
    </StaticConsumer>
  </div>
</Provider>
`;

exports[`Context legacy renders within parent and inherits the defined value 1`] = `
<ParentContext theme="dark">
  <div>
    <ChildContext spy={mockConstructor()}>
      <div>Consumed!</div>
    </ChildContext>
  </div>
</ParentContext>
`;

exports[`Context useContext renders within provider and inherits the defined value 1`] = `
<Provider theme="dark">
  <div>
    <HookConsumer spy={mockConstructor()}>
      <div>Consumed!</div>
    </HookConsumer>
  </div>
</Provider>
`;
